kratos/tests/strategies/strategies/test_strategies.cpp

```diff
@@ -0,0 +1,217 @@
+        typedef DirectSolver<SparseSpaceType,  LocalSpaceType, ReordererType > DirectSolverType;
+        typedef LinearSolver<SparseSpaceType,LocalSpaceType> LinearSolverType;
+        typedef SkylineLUFactorizationSolver<SparseSpaceType,  LocalSpaceType, ReordererType > SkylineLUFactorizationSolverType;
+        
+        // The convergence criteria type
+        typedef ConvergenceCriteria< SparseSpaceType, LocalSpaceType > ConvergenceCriteriaType;
+        typedef ResidualCriteria< SparseSpaceType, LocalSpaceType > ResidualCriteriaType;
+        
+        // The builder ans solver type
+        typedef BuilderAndSolver< SparseSpaceType, LocalSpaceType, LinearSolverType > BuilderAndSolverType;
+        typedef ResidualBasedBlockBuilderAndSolver< SparseSpaceType, LocalSpaceType, LinearSolverType > ResidualBasedBlockBuilderAndSolverType;
+        
+        // The time scheme
+        typedef Scheme< SparseSpaceType, LocalSpaceType >  SchemeType;
+        typedef ResidualBasedIncrementalUpdateStaticScheme< SparseSpaceType, LocalSpaceType> ResidualBasedIncrementalUpdateStaticSchemeType;
+        
+        // The strategies
+        typedef SolvingStrategy<SparseSpaceType, LocalSpaceType, LinearSolverType> SolvingStrategyType;
+        typedef ResidualBasedLinearStrategy< SparseSpaceType, LocalSpaceType, LinearSolverType > ResidualBasedLinearStrategyType;
+        typedef ResidualBasedNewtonRaphsonStrategy< SparseSpaceType, LocalSpaceType, LinearSolverType > ResidualBasedNewtonRaphsonStrategyType;
+        
+        // Dof arrays
+        typedef PointerVectorSet<Dof<double>, SetIdentityFunction<Dof<double>>, std::less<SetIdentityFunction<Dof<double>>::result_type>, std::equal_to<SetIdentityFunction<Dof<double>>::result_type>, Dof<double>* > DofsArrayType;
+        
+        typedef TestElement::ResidualType ResidualType;
+        
+        static inline void BasicTestStrategyDisplacement(
+            ModelPart& ModelPart,
+            const ResidualType ThisResidualType
+            )
+        {
+            ModelPart.SetBufferSize(3);
+            
+            ModelPart.AddNodalSolutionStepVariable(DISPLACEMENT);
+            ModelPart.AddNodalSolutionStepVariable(VELOCITY);
+            ModelPart.AddNodalSolutionStepVariable(ACCELERATION);
+            
+            NodeType::Pointer pnode = ModelPart.CreateNewNode(1, 0.0, 0.0, 0.0);
+            std::vector<NodeType::Pointer> geom(1);
+            geom[0] = pnode;
+            GeometryType::Pointer pgeom = Kratos::make_shared<GeometryType>(geom);
+            Element::Pointer pelem = Kratos::make_shared<TestElement>(1, pgeom, ThisResidualType);
+            ModelPart.AddElement(pelem);
+            
+            pnode->AddDof(DISPLACEMENT_X);
+            pnode->AddDof(DISPLACEMENT_Y);
+            pnode->AddDof(DISPLACEMENT_Z);
+            
+            // Set initial solution
+            const array_1d<double, 3> zero_vector = ZeroVector(3);
```

Comment:
one minor suggestion (you don't need to change the code, just beware of the comment)

if you define an array zero_vector and then assign it, you are doing an allocation (it is stack so not a big deal).

however if you use ZeroVector(3) (and noalias, which is missing in the following line) then no allocation at all happens, as zero vector essentially only defines the operator access to the component you may want. 

the bottomline of my reasoning is: use ZeroVector(size) directly whenever you can.
note also that you may alternatively write
   
      (pnode->FastGetSolutionStepValue(DISPLACEMENT)).clear();

Reply:
Ok, I will do